File: swift-mls/Sources/SwiftMLS/GroupStateUpdate.swift
import CryptoKit
import Foundation

// MARK: - Group State Update Protocol Messages

/// Distributed via the encrypted group channel (kind 24114) after a membership change.
/// Contains the new epoch, salt, and member delta so all members can update their local state.
public struct SEPGroupStateUpdate: Codable, Equatable, Sendable {
    public static let messageType = "sep_state_update"

    public let type: String
    public let epoch: UInt64
    public let salt: Data
    public let addedMembers: [SEPGroupMemberLeaf]
    public let removedMemberKeys: [Data]   // compressed G1 pubkeys of removed members
    public let commitment: Data?           // new Poseidon commitment (optional)
    public let senderAttestation: SEPKeyAttestationPayload?
    public let senderTransportBundle: SEPMemberTransportBundle?

    public init(
        epoch: UInt64,
        salt: Data,
        addedMembers: [SEPGroupMemberLeaf] = [],
        removedMemberKeys: [Data] = [],
        commitment: Data? = nil,
        senderAttestation: SEPKeyAttestationPayload? = nil,
        senderTransportBundle: SEPMemberTransportBundle? = nil
    ) {
        self.type = Self.messageType
        self.epoch = epoch
        self.salt = salt
        self.addedMembers = addedMembers
        self.removedMemberKeys = removedMemberKeys
        self.commitment = commitment
        self.senderAttestation = senderAttestation
        self.senderTransportBundle = senderTransportBundle
    }
}

/// Request a salt for a specific epoch from other online group members.
/// Sent via the encrypted group channel when a member discovers it missed an epoch.
public struct SEPSaltRequest: Codable, Equatable, Sendable {
    public static let messageType = "sep_salt_request"

    public let type: String
    public let epoch: UInt64

    public init(epoch: UInt64) {
        self.type = Self.messageType
        self.epoch = epoch
    }
}

/// Response to a salt request, providing the salt for the requested epoch.
public struct SEPSaltResponse: Codable, Equatable, Sendable {
    public static let messageType = "sep_salt_response"

    public let type: String
    public let epoch: UInt64
    public let salt: Data

    public init(epoch: UInt64, salt: Data) {
        self.type = Self.messageType
        self.epoch = epoch
        self.salt = salt
    }
}

/// A key attestation payload suitable for wire transmission (Codable).
/// Binds a BLS12-381 group membership key to a Stellar Ed25519 account key.
public struct SEPKeyAttestationPayload: Codable, Equatable, Sendable {
    public let blsPubkey: Data       // 48 bytes, compressed G1 point
    public let ed25519Pubkey: Data   // 32 bytes, Stellar Ed25519 public key
    public let signature: Data       // 64 bytes, Ed25519 signature

    public init(blsPubkey: Data, ed25519Pubkey: Data, signature: Data) {
        self.blsPubkey = blsPubkey
        self.ed25519Pubkey = ed25519Pubkey
        self.signature = signature
    }

    /// Validate the structural integrity of the attestation fields.
    /// Returns false if any field has an incorrect byte length.
    public var hasValidStructure: Bool {
        blsPubkey.count == 48 && ed25519Pubkey.count == 32 && signature.count == 64
    }

    /// Compute the binding message that the Ed25519 signature covers:
    /// `SHA-256("SEP-XXXX:key-binding" || bls_pubkey)`
    ///
    /// The caller MUST verify `signature` over this message using
    /// `ed25519Pubkey` via CryptoKit or equivalent. The SDK does not
    /// perform the Ed25519 verification itself.
    public func computeBindingMessage() -> Data {
        var hasher = SHA256()
        hasher.update(data: Data("SEP-XXXX:key-binding".utf8))
        hasher.update(data: blsPubkey)
        return Data(hasher.finalize())
    }
}

/// Broadcast by a new joiner after joining via invite code.
/// Existing members receive this, add the joiner to their member list,
/// and respond with a `SEPGroupStateUpdate` containing the updated state.
public struct SEPMemberJoined: Codable, Equatable, Sendable {
    public static let messageType = "sep_member_joined"

    public let type: String
    public let member: SEPGroupMemberLeaf
    public let transportBundle: SEPMemberTransportBundle?

    public init(member: SEPGroupMemberLeaf, transportBundle: SEPMemberTransportBundle? = nil) {
        self.type = Self.messageType
        self.member = member
        self.transportBundle = transportBundle
    }
}

/// Broadcast when a member renames the group.
public struct SEPGroupRenamed: Codable, Equatable, Sendable {
    public static let messageType = "sep_group_renamed"

    public let type: String
    public let name: String

    public init(name: String) {
        self.type = Self.messageType
        self.name = name
    }
}

/// Post-removal re-key message. Sent immediately after a member removal state update.
/// Contains the real salt (the state update carries a poisoned/placeholder salt so the
/// removed member cannot derive the new encryption key). Encrypted with the key derived
/// from the poisoned salt — only members still subscribed will receive and process it.
public struct SEPRekey: Codable, Equatable, Sendable {
    public static let messageType = "sep_rekey"

    public let type: String
    public let epoch: UInt64
    public let salt: Data

    public init(epoch: UInt64, salt: Data) {
        self.type = Self.messageType
        self.epoch = epoch
        self.salt = salt
    }
}

/// Delivery acknowledgment sent when a member receives and decrypts a message.
/// Each device sends at most one ACK per original event ID.
public struct SEPMessageAck: Codable, Equatable, Sendable {
    public static let messageType = "sep_message_ack"

    public let type: String
    public let eventID: String   // the original Nostr event ID that was delivered

    public init(eventID: String) {
        self.type = Self.messageType
        self.eventID = eventID
    }
}

// MARK: - Authenticated Member Transport Bundle

/// Authenticated mapping from a group member to their X25519 inbox encryption key.
/// Enables per-member asymmetric delivery of rekey material during removals.
/// The signature covers a domain-separated hash of all fields, signed by the member's
/// Stellar Ed25519 key, ensuring the inbox key cannot be spoofed.
public struct SEPMemberTransportBundle: Codable, Equatable, Sendable {
    public static let currentVersion: UInt16 = 1

    public let blsPubkey: Data             // 48 bytes, compressed G1
    public let stellarEd25519Pubkey: Data   // 32 bytes
    public let x25519InboxPubkey: Data      // 32 bytes
    public let version: UInt16
    public let signature: Data              // 64 bytes, Ed25519

    public init(
        blsPubkey: Data,
        stellarEd25519Pubkey: Data,
        x25519InboxPubkey: Data,
        version: UInt16 = currentVersion,
        signature: Data
    ) {
        self.blsPubkey = blsPubkey
        self.stellarEd25519Pubkey = stellarEd25519Pubkey
        self.x25519InboxPubkey = x25519InboxPubkey
        self.version = version
        self.signature = signature
    }

    public var hasValidStructure: Bool {
        blsPubkey.count == 48 &&
        stellarEd25519Pubkey.count == 32 &&
        x25519InboxPubkey.count == 32 &&
        signature.count == 64
    }

    /// Domain-separated binding message:
    /// SHA-256("SEP-XXXX:member-transport-v1" || bls || stellar || x25519 || version)
    public func computeBindingMessage() -> Data {
        var hasher = CryptoKit.SHA256()
        hasher.update(data: Data("SEP-XXXX:member-transport-v1".utf8))
        hasher.update(data: blsPubkey)
        hasher.update(data: stellarEd25519Pubkey)
        hasher.update(data: x25519InboxPubkey)
        var vBytes = version.bigEndian
        withUnsafeBytes(of: &vBytes) { hasher.update(bufferPointer: $0) }
        return Data(hasher.finalize())
    }
}

/// Non-secret notice broadcast on the old group channel during a secure removal.
/// Does NOT contain new groupSecret or salt — only metadata.
public struct SEPRemovalNotice: Codable, Equatable, Sendable {
    public static let messageType = "sep_removal_notice"

    public let type: String
    public let groupID: Data
    public let epoch: UInt64
    public let removedMemberKeys: [Data]
    public let commitment: Data?

    public init(
        groupID: Data,
        epoch: UInt64,
        removedMemberKeys: [Data],
        commitment: Data? = nil
    ) {
        self.type = Self.messageType
        self.groupID = groupID
        self.epoch = epoch
        self.removedMemberKeys = removedMemberKeys
        self.commitment = commitment
    }
}

/// Per-member rekey envelope delivered via X25519 inbox transport.
/// Contains fresh groupSecret and salt that the removed member never receives.
public struct SEPRekeyEnvelope: Codable, Equatable, Sendable {
    public static let messageType = "sep_rekey_envelope"

    public let type: String
    public let groupID: Data              // 32 bytes
    public let epoch: UInt64
    public let groupSecret: Data          // 32 bytes — the NEW group secret
    public let salt: Data                 // 32 bytes — the NEW salt
    public let commitment: Data?
    public let memberBundles: [SEPMemberTransportBundle]
    public let removedMemberKeys: [Data]
    public let relayHints: [String]
    public let senderBundle: SEPMemberTransportBundle

    public init(
        groupID: Data,
        epoch: UInt64,
        groupSecret: Data,
        salt: Data,
        commitment: Data? = nil,
        memberBundles: [SEPMemberTransportBundle],
        removedMemberKeys: [Data],
        relayHints: [String],
        senderBundle: SEPMemberTransportBundle
    ) {
        self.type = Self.messageType
        self.groupID = groupID
        self.epoch = epoch
        self.groupSecret = groupSecret
        self.salt = salt
        self.commitment = commitment
        self.memberBundles = memberBundles
        self.removedMemberKeys = removedMemberKeys
        self.relayHints = relayHints
        self.senderBundle = senderBundle
    }
}

/// Acknowledgment that a member received and installed a rekey envelope.
public struct SEPRekeyAck: Codable, Equatable, Sendable {
    public static let messageType = "sep_rekey_ack"

    public let type: String
    public let groupID: Data
    public let epoch: UInt64
    public let senderBlsPubkey: Data

    public init(groupID: Data, epoch: UInt64, senderBlsPubkey: Data) {
        self.type = Self.messageType
        self.groupID = groupID
        self.epoch = epoch
        self.senderBlsPubkey = senderBlsPubkey
    }
}

/// Request for a rekey envelope resend (sent via inbox when a member misses the original).
public struct SEPRekeyResendRequest: Codable, Equatable, Sendable {
    public static let messageType = "sep_rekey_resend_request"

    public let type: String
    public let groupID: Data
    public let epoch: UInt64
    public let requesterBundle: SEPMemberTransportBundle

    public init(groupID: Data, epoch: UInt64, requesterBundle: SEPMemberTransportBundle) {
        self.type = Self.messageType
        self.groupID = groupID
        self.epoch = epoch
        self.requesterBundle = requesterBundle
    }
}

// MARK: - Protocol Message Envelope

/// Wrapper for protocol messages sent over the encrypted group channel.
/// Allows distinguishing between plain text chat messages and protocol messages.
public struct SEPProtocolMessage: Codable, Sendable {
    public let type: String

    /// Attempt to parse a decrypted message as a protocol message.
    /// Returns nil if the message is plain text (not a protocol message).
    public static func parse(_ json: String) -> String? {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONDecoder().decode(SEPProtocolMessage.self, from: data)
        else { return nil }
        return obj.type
    }
}

// MARK: - Relayer Configuration

/// Configuration for fee-decoupled transaction submission via a relayer.
///
/// The relayer receives the same contract invocation payload but wraps it
/// in a Stellar transaction signed by its own key, paying the network fee
/// on behalf of the group member. This prevents identity leakage through
/// the transaction signer address.
public struct SEPRelayerConfig: Codable, Equatable, Sendable {
    public let relayerURL: URL
    public let authToken: String?
    /// SHA-256 hashes of the relayer's TLS certificate public keys (base64-encoded).
    /// When non-empty, the transport will reject connections whose server certificate
    /// does not match any of the pinned hashes (H-14: TLS certificate pinning).
    public let pinnedCertificateHashes: [String]

    public init(relayerURL: URL, authToken: String? = nil, pinnedCertificateHashes: [String] = []) {
        self.relayerURL = relayerURL
        self.authToken = authToken
        self.pinnedCertificateHashes = pinnedCertificateHashes
    }
}

File: kotlin-mls/src/main/java/com/stellarmls/mls/GroupStateUpdate.kt
package com.stellarmls.mls

// MARK: - Protocol Message Type Registry
//
// CHECKLIST: When adding a new protocol message type:
// 1. Add it to SEPMessageType below
// 2. Define the data class in this file
// 3. Add to isProtocolMessage on both platforms (iOS NostrMessageTransport, Android NostrMessageTransport)
// 4. If it modifies membership: add to applyMemberChanges on BOTH platforms
// 5. Add dispatch case in the main protocol message handler on both platforms
// 6. Add cross-platform test vector in docs/cross-platform-test-vectors.json

/**
 * Exhaustive registry of all SEP protocol message types.
 * Every MESSAGE_TYPE constant in this file MUST have a corresponding entry here.
 */
enum class SEPMessageType(val wireValue: String, val modifiesMembership: Boolean) {
    STATE_UPDATE("sep_state_update", modifiesMembership = true),
    MEMBER_JOINED("sep_member_joined", modifiesMembership = true),
    SALT_REQUEST("sep_salt_request", modifiesMembership = false),
    SALT_RESPONSE("sep_salt_response", modifiesMembership = false),
    GROUP_RENAMED("sep_group_renamed", modifiesMembership = false),
    REKEY("sep_rekey", modifiesMembership = false),
    MESSAGE_ACK("sep_message_ack", modifiesMembership = false),
    REMOVAL_NOTICE("sep_removal_notice", modifiesMembership = true),
    REKEY_ENVELOPE("sep_rekey_envelope", modifiesMembership = false),
    REKEY_ACK("sep_rekey_ack", modifiesMembership = false),
    REKEY_RESEND_REQUEST("sep_rekey_resend_request", modifiesMembership = false);

    companion object {
        private val byWireValue = entries.associateBy { it.wireValue }
        fun fromWireValue(value: String): SEPMessageType? = byWireValue[value]

        /** All message types that require applyMemberChanges handling. */
        val membershipModifying: Set<SEPMessageType> = entries.filter { it.modifiesMembership }.toSet()
    }
}

// MARK: - Group State Update Protocol Messages

/**
 * Distributed via the encrypted group channel (kind 24114) after a membership change.
 * Contains the new epoch, salt, and member delta so all members can update their local state.
 */
data class SEPGroupStateUpdate(
    val type: String = MESSAGE_TYPE,
    val epoch: Long,
    val salt: ByteArray,
    val addedMembers: List<SEPGroupMemberLeaf> = emptyList(),
    val removedMemberKeys: List<ByteArray> = emptyList(),
    val commitment: ByteArray? = null,
    val senderAttestation: SEPKeyAttestationPayload? = null,
    val senderTransportBundle: SEPMemberTransportBundle? = null,
    /** All known transport bundles for group members. Enables secure rekey
     *  by distributing bundles to members who joined after others. */
    val memberBundles: List<SEPMemberTransportBundle> = emptyList()
) {
    companion object {
        const val MESSAGE_TYPE = "sep_state_update"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPGroupStateUpdate) return false
        return epoch == other.epoch && salt.contentEquals(other.salt)
    }

    override fun hashCode(): Int = 31 * epoch.hashCode() + salt.contentHashCode()
}

/**
 * Broadcast by a new joiner after joining via invite code.
 * Existing members receive this, add the joiner to their member list,
 * and respond with an [SEPGroupStateUpdate] containing the updated state.
 */
data class SEPMemberJoined(
    val type: String = MESSAGE_TYPE,
    val member: SEPGroupMemberLeaf,
    val transportBundle: SEPMemberTransportBundle? = null
) {
    companion object {
        const val MESSAGE_TYPE = "sep_member_joined"
    }
}

/**
 * Request a salt for a specific epoch from other online group members.
 * Sent via the encrypted group channel when a member discovers it missed an epoch.
 */
data class SEPSaltRequest(
    val type: String = MESSAGE_TYPE,
    val epoch: Long
) {
    companion object {
        const val MESSAGE_TYPE = "sep_salt_request"
    }
}

/**
 * Response to a salt request, providing the salt for the requested epoch.
 */
data class SEPSaltResponse(
    val type: String = MESSAGE_TYPE,
    val epoch: Long,
    val salt: ByteArray
) {
    companion object {
        const val MESSAGE_TYPE = "sep_salt_response"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPSaltResponse) return false
        return epoch == other.epoch && salt.contentEquals(other.salt)
    }

    override fun hashCode(): Int = 31 * epoch.hashCode() + salt.contentHashCode()
}

/**
 * A key attestation payload suitable for wire transmission.
 * Binds a BLS12-381 group membership key to a Stellar Ed25519 account key.
 */
data class SEPKeyAttestationPayload(
    val blsPubkey: ByteArray,       // 48 bytes, compressed G1 point
    val ed25519Pubkey: ByteArray,   // 32 bytes, Stellar Ed25519 public key
    val signature: ByteArray        // 64 bytes, Ed25519 signature
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPKeyAttestationPayload) return false
        return blsPubkey.contentEquals(other.blsPubkey) &&
                ed25519Pubkey.contentEquals(other.ed25519Pubkey) &&
                signature.contentEquals(other.signature)
    }

    override fun hashCode(): Int {
        var result = blsPubkey.contentHashCode()
        result = 31 * result + ed25519Pubkey.contentHashCode()
        result = 31 * result + signature.contentHashCode()
        return result
    }
}

/**
 * Post-removal re-key message. Sent immediately after a member removal state update.
 * Contains the real salt (the state update carries a poisoned/placeholder salt so the
 * removed member cannot derive the new encryption key). Encrypted with the key derived
 * from the poisoned salt — only members still subscribed will receive and process it.
 */
data class SEPRekey(
    val type: String = MESSAGE_TYPE,
    val epoch: Long,
    val salt: ByteArray
) {
    companion object {
        const val MESSAGE_TYPE = "sep_rekey"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPRekey) return false
        return epoch == other.epoch && salt.contentEquals(other.salt)
    }

    override fun hashCode(): Int = 31 * epoch.hashCode() + salt.contentHashCode()
}

/**
 * Delivery acknowledgment sent when a member receives and decrypts a message.
 * Each device sends at most one ACK per original event ID.
 */
data class SEPMessageAck(
    val type: String = MESSAGE_TYPE,
    val eventID: String
) {
    companion object {
        const val MESSAGE_TYPE = "sep_message_ack"
    }
}

/**
 * Broadcast when a member renames the group.
 */
data class SEPGroupRenamed(
    val type: String = MESSAGE_TYPE,
    val name: String
) {
    companion object {
        const val MESSAGE_TYPE = "sep_group_renamed"
    }
}

// MARK: - Authenticated Member Transport Bundle

/**
 * Authenticated mapping from a group member to their X25519 inbox encryption key.
 * Enables per-member asymmetric delivery of rekey material during removals.
 * The signature covers a domain-separated hash of all fields, signed by the member's
 * Stellar Ed25519 key, ensuring the inbox key cannot be spoofed.
 */
data class SEPMemberTransportBundle(
    val blsPubkey: ByteArray,            // 48 bytes, compressed G1
    val stellarEd25519Pubkey: ByteArray, // 32 bytes
    val x25519InboxPubkey: ByteArray,    // 32 bytes
    val version: Int = CURRENT_VERSION,  // u16
    val signature: ByteArray             // 64 bytes, Ed25519
) {
    companion object {
        const val CURRENT_VERSION = 1
    }

    val hasValidStructure: Boolean
        get() = blsPubkey.size == 48 &&
                stellarEd25519Pubkey.size == 32 &&
                x25519InboxPubkey.size == 32 &&
                signature.size == 64

    /** Domain-separated binding message:
     *  SHA-256("SEP-XXXX:member-transport-v1" || bls || stellar || x25519 || version) */
    fun computeBindingMessage(): ByteArray {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        digest.update("SEP-XXXX:member-transport-v1".toByteArray())
        digest.update(blsPubkey)
        digest.update(stellarEd25519Pubkey)
        digest.update(x25519InboxPubkey)
        digest.update(byteArrayOf((version shr 8).toByte(), version.toByte()))
        return digest.digest()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPMemberTransportBundle) return false
        return blsPubkey.contentEquals(other.blsPubkey) &&
                stellarEd25519Pubkey.contentEquals(other.stellarEd25519Pubkey) &&
                x25519InboxPubkey.contentEquals(other.x25519InboxPubkey) &&
                version == other.version &&
                signature.contentEquals(other.signature)
    }

    override fun hashCode(): Int {
        var result = blsPubkey.contentHashCode()
        result = 31 * result + stellarEd25519Pubkey.contentHashCode()
        result = 31 * result + x25519InboxPubkey.contentHashCode()
        result = 31 * result + version
        result = 31 * result + signature.contentHashCode()
        return result
    }
}

/**
 * Non-secret notice broadcast on the old group channel during a secure removal.
 * Does NOT contain new groupSecret or salt — only metadata.
 */
data class SEPRemovalNotice(
    val type: String = MESSAGE_TYPE,
    val groupID: ByteArray,
    val epoch: Long,
    val removedMemberKeys: List<ByteArray>,
    val commitment: ByteArray? = null
) {
    companion object {
        const val MESSAGE_TYPE = "sep_removal_notice"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPRemovalNotice) return false
        return groupID.contentEquals(other.groupID) && epoch == other.epoch
    }

    override fun hashCode(): Int = 31 * groupID.contentHashCode() + epoch.hashCode()
}

/**
 * Per-member rekey envelope delivered via X25519 inbox transport.
 * Contains fresh groupSecret and salt that the removed member never receives.
 */
data class SEPRekeyEnvelope(
    val type: String = MESSAGE_TYPE,
    val groupID: ByteArray,           // 32 bytes
    val epoch: Long,
    val groupSecret: ByteArray,       // 32 bytes — the NEW group secret
    val salt: ByteArray,              // 32 bytes — the NEW salt
    val commitment: ByteArray? = null,
    val memberBundles: List<SEPMemberTransportBundle>,
    val removedMemberKeys: List<ByteArray>,
    val relayHints: List<String>,
    val senderBundle: SEPMemberTransportBundle
) {
    companion object {
        const val MESSAGE_TYPE = "sep_rekey_envelope"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPRekeyEnvelope) return false
        return groupID.contentEquals(other.groupID) && epoch == other.epoch
    }

    override fun hashCode(): Int = 31 * groupID.contentHashCode() + epoch.hashCode()
}

/**
 * Acknowledgment that a member received and installed a rekey envelope.
 */
data class SEPRekeyAck(
    val type: String = MESSAGE_TYPE,
    val groupID: ByteArray,
    val epoch: Long,
    val senderBlsPubkey: ByteArray
) {
    companion object {
        const val MESSAGE_TYPE = "sep_rekey_ack"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPRekeyAck) return false
        return groupID.contentEquals(other.groupID) && epoch == other.epoch
    }

    override fun hashCode(): Int = 31 * groupID.contentHashCode() + epoch.hashCode()
}

/**
 * Request for a rekey envelope resend (sent via inbox when a member misses the original).
 */
data class SEPRekeyResendRequest(
    val type: String = MESSAGE_TYPE,
    val groupID: ByteArray,
    val epoch: Long,
    val requesterBundle: SEPMemberTransportBundle
) {
    companion object {
        const val MESSAGE_TYPE = "sep_rekey_resend_request"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SEPRekeyResendRequest) return false
        return groupID.contentEquals(other.groupID) && epoch == other.epoch
    }

    override fun hashCode(): Int = 31 * groupID.contentHashCode() + epoch.hashCode()
}

/**
 * Configuration for fee-decoupled transaction submission via a relayer.
 *
 * The relayer receives the same contract invocation payload but wraps it
 * in a Stellar transaction signed by its own key, paying the network fee.
 */
data class SEPRelayerConfig(
    val relayerURL: String,
    val authToken: String? = null
)
